paasta_tools/apply_external_resources.py

```diff
@@ -0,0 +1,76 @@
+#!/opt/venvs/paasta-tools/bin/python
+import os
+import sys
+from filecmp import cmp
+from shutil import copy
+from subprocess import CalledProcessError
+from subprocess import run
+from traceback import print_exc
+
+APPLIED_DIRECTORY = ".applied"
+
+
+# This script expects the KUBECONFIG environment variable to be set correctly
+def main(puppet_resource_root: str) -> int:
+    exit_code = 0
+    applied_resource_root = os.path.join(puppet_resource_root, APPLIED_DIRECTORY)
+
+    # Loop through everything in the puppet resource path
+    for root, dirs, files in os.walk(puppet_resource_root):
+        # modifying the 'dirs' variable in-place will update the order that os.walk visits things
+        if APPLIED_DIRECTORY in dirs:
+            dirs.remove(APPLIED_DIRECTORY)
+        dirs.sort()  # Need to apply things in the correct order
```

Comment:
Please correct me if I'm wrong, but it doesn't seem that `dirs` variable is used anywhere?

Reply:
`dirs` is used by `os.walk` -- if you modify it in place, it will change the order in which directories are visited:

> When topdown is True, the caller can modify the dirnames list in-place (perhaps using del or slice assignment), and walk() will only recurse into the subdirectories whose names remain in dirnames; this can be used to prune the search, impose a specific order of visiting, or even to inform walk() about directories the caller creates or renames before it resumes walk() again.

(from the Python documentation: https://docs.python.org/3.8/library/os.html#os.walk)

Reply:
Good to know!
Resolving it.